teslajsonpy/homeassistant/homelink.py

```diff
@@ -0,0 +1,90 @@
+#  SPDX-License-Identifier: Apache-2.0
+"""
+Python Package for controlling Tesla API.
+
+For more details about this api, please refer to the documentation at
+https://github.com/zabuldon/teslajsonpy
+"""
+
+from teslajsonpy.exceptions import HomelinkError
+from teslajsonpy.homeassistant.vehicle import VehicleDevice
+
+
+class TriggerHomelink(VehicleDevice):
+    """Home-Assistant class for trigger homelink of Tesla vehicles."""
+
+    def __init__(self, data, controller):
+        """Initialize the trigger homelink for the vehicle.
+
+        Parameters
+        ----------
+        data : dict
+            The trigger homelink for a Tesla vehicle.
+            https://tesla-api.timdorr.com/vehicle/commands/homelink
+        controller : teslajsonpy.Controller
+            The controller that controls updates to the Tesla API.
+
+        Returns
+        -------
+        None
+
+        """
+        super().__init__(data, controller)
+        self.type = "trigger homelink"
+        self.hass_type = "button"
+        self.name = self._name()
+        self.uniq_name = self._uniq_name()
+        self.__longitude = None
+        self.__latitude = None
+        self.__homelink_device_count = None
+        self.__homelink_nearby = None
+        self.__homelink_available = False
+
+    async def async_update(self, wake_if_asleep=False, force=False):
+        """Update the trigger homelink of the vehicle."""
+        await super().async_update(wake_if_asleep=wake_if_asleep)
+        self.refresh()
+
+    @staticmethod
+    def has_battery() -> bool:
+        """Return whether the device has a battery."""
+        return False
+
+    async def trigger_homelink(self) -> None:
+        """Trigger Homelink."""
+        if self.__latitude is not None and self.__longitude is not None:
+            if not self.__homelink_device_count:
+                raise HomelinkError(f"No homelink devices added to {self.car_name()}.")
+            if not self.__homelink_nearby:
+                raise HomelinkError(f"No homelink devices near {self.car_name()}.")
+            await self._controller.api(
+                "TRIGGER_HOMELINK",
+                path_vars={"vehicle_id": self._id},
+                lat=self.__latitude,
+                lon=self.__longitude,
+                wake_if_asleep=True,
+            )
+
+    def refresh(self) -> None:
+        """Refresh data.
+
+        This assumes the controller has already been updated
+        """
+        super().refresh()
+        data = self._controller.get_drive_params(self._id)
+        if data:
+            if data["native_location_supported"]:
```

Comment:
Don't forget to test both branches.

Reply:
Is it acceptable to reference "private" values in a test? There's no method which returns the lan/lon, so to test this I'd need to assert it's .__latitude and .__longitude are as expected from the test, which seems wrong.

Or is it better to add a method in TriggerHomelink specifically to return these for testing?

Reply:
I partially answered this, with double underscore I can't access them, so I could change to single leading underscore to make them not really private to test, or add a method to return them specifically for testing, whichever you prefer.

Reply:
You can control the value of data. In this case, make the return false.

Reply:
Thanks, I still wasn't sure of best approach so I took a stab with [a81c574](https://github.com/zabuldon/teslajsonpy/pull/296/commits/a81c574ddee1187fe28e98ae32bb39e4875f331f) by changing the private vars to a single leading underscore so the test can assert them, let me know if that works.

Reply:
It's better practice not to have to change private variables. You should be able to mock the data info in the test instead. Please reference other tests. (I haven't reviewed the tests here for a while, so if you look and everyone else has just had single underscore variables, then I guess you can follow that trend).

Reply:
Sorry I think we were talking about 2 different things:
I'm not setting private variables, I just needed to read the private variables to assert them in test. Since python 3.10 doesn't allow accessing double underscore private variables I had to rename them to single leading underscore so they can be read in the test.
This is why I was asking is it better to make a new method in TriggerHomelink to return the private values instead of reading them directly from the test. Both options are not great: one is reading private variables, the other is making a method which is never used except in tests, so I wanted to ask which is the least-worst option.

As the test is currently, it's mocking the data, setting the controller to mocked data, updating the button, and ensuring the button's (private) lat/lon values are correct. Is this ok?
```python3
    _data = _mock.data_request_vehicle()
    _data["drive_state"]["native_location_supported"] = 1
    _data["drive_state"]["longitude"] = 12.345
    _data["drive_state"]["native_longitude"] = 23.456
    _data["drive_state"]["latitude"] = 34.567
    _data["drive_state"]["native_latitude"] = 45.678
    _button = TriggerHomelink(_data, _controller)

    _controller.set_drive_params(vin=VIN, params=_data["drive_state"])

    await _button.async_update()

    assert _button._longitude == 23.456
    assert _button._latitude == 45.678
```
(and of course the non-native version was added as well)

Reply:
Oh sorry.  I guess, the cleaner method would be to have an access function to get all the internal state. While no one is accessing the button state now, it'd be more extensible in the future instead of encouraging protected variable direct access. But honestly, you're right that seems wasteful since it's caused by needing to test. Isn't it just a warning if you try to directly access a private variable during a test? Perhaps we can squelch the warning. I'm ok with what you think makes the most sense in this case. 

I guess the weirdness for me was the fact that this is caching internal state about the vehicle instead of accessing the data directly from the controller when needed.

Reply:
> Oh sorry. I guess, the cleaner method would be to have an access function to get all the internal state. While no one is accessing the button state now, it'd be more extensible in the future instead of encouraging protected variable direct access. But honestly, you're right that seems wasteful since it's caused by needing to test. Isn't it just a warning if you try to directly access a private variable during a test? Perhaps we can squelch the warning. I'm ok with what you think makes the most sense in this case.

Now that it's `_latitude` there's no warning, but originally it was `__latitude` which raised a value error that the variable didn't exist, since it's not accessible at all. 

> 
> I guess the weirdness for me was the fact that this is caching internal state about the vehicle instead of accessing the data directly from the controller when needed.

ahh good point, it doesn't actually need to be cached, but if it's not cached, then I really don't see how to test that it's using the right location in the api call. Apologies for my ignorance as I'm brand new to mocks, but is there a way to intercept the api call to check the parameters? I haven't found any other entity tests here to do that, nor did I find anything like that in TeslaMock.

Reply:
Take a look at patch. https://docs.python.org/3/library/unittest.mock.html#quick-guide

Reply:
Oh sorry I missed the part about checking parameters. You may be overthinking this. You just need to test that it doesn't throw an exception for being too far when the data includes native locations or not. That will test the two paths. I don't think you have to test the actual internal state.

There may be a way to do what you're proposing with checking the state, but really the goal is to make sure that someone doesn't break your code later and I think black box testing is fine.

Reply:
Unfortunately in unit testing, the lat/lon values have no affect at all, so there's no way to determine if it would throw an exception for being too far, as that's entirely determined by the car itself.

I took a stab at patching controller.api so I could assert_called_with to ensure we're calling with the right lat/lon, however I hit a road block that monkeypatch doesn't offer this feature, and it seems implementing mock/magicmock would break/overwrite our existing TeslaMock/monkeypatch setup.

I also found no other modules in teslajsonpy doing this type of assertion, as they all only inspect what's internal to the lock/switch/climate classes, and don't assert that it's making the right api calls, so you're right that I'm probably overthinking this haha.

if it's ok with you we'll leave it where it's checking the cached lat/lon values are correct, and trust that the api call is made correctly, as the other modules are doing?